zug/util.hpp

```diff
@@ -89,29 +90,69 @@ constexpr decltype(auto) invoke_composition(Fns&& fns, Args&&... args)
 }
 
 template <typename Fn, typename... Fns>
-struct composed
+struct composed : std::tuple<Fn, Fns...>
 {
-    std::tuple<Fn, Fns...> fns;
+    using function_tuple = std::tuple<Fn, Fns...>;
+
+    template <typename ArgsTuple>
+    composed(ArgsTuple&& args)
+        : function_tuple{std::forward<ArgsTuple>(args)}
+    {}
 
     template <typename... T>
     constexpr decltype(auto) operator()(T&&... xs) &
     {
-        return invoke_composition(fns, std::forward<T>(xs)...);
+        return invoke_composition(static_cast<function_tuple&>(*this),
+                                  std::forward<T>(xs)...);
     }
 
     template <typename... T>
     constexpr decltype(auto) operator()(T&&... xs) const&
     {
-        return invoke_composition(fns, std::forward<T>(xs)...);
+        return invoke_composition(static_cast<const function_tuple&>(*this),
+                                  std::forward<T>(xs)...);
     }
 
     template <typename... T>
     constexpr decltype(auto) operator()(T&&... xs) &&
     {
-        return invoke_composition(std::move(fns), std::forward<T>(xs)...);
+        return invoke_composition(
+            std::move(static_cast<function_tuple&>(*this)),
+            std::forward<T>(xs)...);
     }
 };
 
+template <typename T>
+constexpr bool is_composed_v = false;
+
+template <typename... Fns>
+constexpr bool is_composed_v<composed<Fns...>> = true;
+
+template <typename Composed,
+          std::enable_if_t<is_composed_v<Composed>>* = nullptr>
+decltype(auto) to_function_tuple(Composed&& c, meta::try_t)
+{
+    return std::forward<typename Composed::function_tuple>(c);
```

Comment:
I do not think the forwarding is correct here, because `typename Composed::function_tuple` is always a value type and does not keep the reference categories of `Composed`. One option is to add a method `as_tuple()` to `composed<>` with the three `&`, `const&`, `&&` overloads, then here you do `std::forward<Composed>(c).as_tuple()`. Another option is to use some `meta` tool I made `copy_decay`, but I think I removed it at some point. The advantage of the first approach is also that you can use it in the `operator()` to get rid of the `static_cast`s that you have now.

Reply:
OTOH I do not think you really need to remove the `composed<>` out of the type, I mean, you can just return `std::forward<Composed>(c)`. Same above, I do not think the `static_cast`s are reaaally needed.

Reply:
clang was complaining about tuple_cat and tuple_size with something that wasn’t exactly a std::tuple

Reply:
(Also: duh 🤦)